sklearn/partial_dependence.py

```diff
@@ -0,0 +1,574 @@
+        The method to use to calculate the partial dependence function:
+
+        - If 'recursion', the underlying trees of ``est`` will be recursed to
+          calculate the function. Only supported for BaseGradientBoosting and
+          ForestRegressor.
+        - If 'exact', the function will be calculated by calling the
+          ``predict_proba`` method of ``est`` for classification or ``predict``
+          for regression on ``X``for every point in the grid. To speed up this
+          method, you can use a subset of ``X`` or a more coarse grid.
+        - If 'estimated', the function will be calculated by calling the
+          ``predict_proba`` method of ``est`` for classification or ``predict``
+          for regression on the mean of ``X``.
+        - If None, then 'recursion' will be used if ``est`` is
+          BaseGradientBoosting or ForestRegressor, and 'exact' used for other
+          estimators.
+
+    Returns
+    -------
+    pdp : array, shape=(n_classes, n_points)
+        The partial dependence function evaluated on the ``grid``.
+        For regression and binary classification ``n_classes==1``.
+    axes : seq of ndarray or None
+        The axes with which the grid has been created or None if
+        the grid has been given.
+
+    Examples
+    --------
+    >>> samples = [[0, 0, 2], [1, 0, 0]]
+    >>> labels = [0, 1]
+    >>> from sklearn.ensemble import GradientBoostingClassifier
+    >>> gb = GradientBoostingClassifier(random_state=0).fit(samples, labels)
+    >>> kwargs = dict(X=samples, percentiles=(0, 1), grid_resolution=2)
+    >>> partial_dependence(gb, [0], **kwargs) # doctest: +SKIP
+    (array([[-4.52...,  4.52...]]), [array([ 0.,  1.])])
+    """
+    if method is None:
+        if isinstance(est, (BaseGradientBoosting, ForestRegressor)):
+            method = 'recursion'
+        else:
+            method = 'exact'
+    if (not isinstance(est, (BaseGradientBoosting, ForestRegressor)) and
+            method == 'recursion'):
+        raise ValueError('est has to be an instance of BaseGradientBoosting or'
+                         ' ForestRegressor for the "recursion" method. Try '
+                         'using method="exact" or "estimated".')
+    if (method != 'recursion' and
+            not hasattr(est, 'predict_proba') and
+            isinstance(est, ClassifierMixin)):
+        raise ValueError('est requires a predict_proba method for '
+                         'method="exact" or "estimated".')
+    if method == 'recursion':
+        if len(est.estimators_) == 0:
+            raise ValueError('Call %s.fit before partial_dependence' %
+                             est.__class__.__name__)
+        if isinstance(est, BaseGradientBoosting):
+            n_features = est.n_features
+        else:
+            n_features = est.n_features_
+    elif X is None:
+        raise ValueError('X is required for method="exact" or "estimated".')
```

Comment:
Why is this so? Seems like you should be able to do exact/estimated computation on the grid too, right?